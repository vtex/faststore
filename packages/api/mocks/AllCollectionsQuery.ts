export const AllCollectionsQueryFirst5 = `query allCollections {
  allCollections(first: 5) {
    edges {
      cursor
      node {
        id
        slug
        type
        breadcrumbList {
          itemListElement {
            item
            name
            position
          }
          numberOfItems
        }
        meta {
          selectedFacets {
            key
            value
          }
        }
        seo {
          canonical
          description
          title
          titleTemplate
        }
      }
    }
    pageInfo {
      endCursor
      hasNextPage
      hasPreviousPage
      startCursor
      totalCount
    }
  }
}
`

export const catalogBrandListFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/brand/list',
  init: undefined,
  result: JSON.parse(
    `[{"id":9280,"name":"Brand","isActive":true,"title":"Brand","metaTagDescription":"Brand","imageUrl":null},{"id":2000001,"name":"Skechers","isActive":true,"title":"Skechers","metaTagDescription":"Sport, casual, work, wide, kids' & performance shoes with style, comfort, innovation, quality & value.","imageUrl":"/155421/skechers.jpg"},{"id":2000002,"name":"Acer","isActive":true,"title":"Acer","metaTagDescription":"Acer laptops, desktops as well as servers and storage, personal digital assistance (PDA), peripherals, peripherals and e-business services for government, business, education, and home users.","imageUrl":null},{"id":2000003,"name":"iRobot","isActive":true,"title":"iRobot","metaTagDescription":"iRobot, the leading global consumer robot company, designs and builds robots that empower people to do more both inside and outside of the home.","imageUrl":null},{"id":2000004,"name":"adidas","isActive":true,"title":"adidas","metaTagDescription":"adidas shoes, clothing and view new collections for adidas Originals, running, football, soccer, training and much more.","imageUrl":null},{"id":2000005,"name":"BLACK+DECKER","isActive":true,"title":"BLACK+DECKER","metaTagDescription":"Stanley Black & Decker, Inc. is a global provider of hand tools, power tools and related accessories, mechanical access solutions, such as automatic doors and commercial locking systems, electronic security and monitoring systems, healthcare solutions, engineered fastening systems, and products and services for various industrial applciations.","imageUrl":null},{"id":2000006,"name":"Nike","isActive":true,"title":"Nike","metaTagDescription":"Nike designs, develops, and sells a variety of products to help in playing basketball and soccer (football), as well as in running, men's and women's training, and other action sports.","imageUrl":null},{"id":2000007,"name":"Sany","isActive":false,"title":"Sany","metaTagDescription":"Camera, Photo","imageUrl":null},{"id":2000008,"name":"Nykor","isActive":false,"title":"Nykor","metaTagDescription":"Camera, Photo","imageUrl":null},{"id":2000009,"name":"Canyon","isActive":false,"title":"Canyon","metaTagDescription":"Camera, Photo","imageUrl":null},{"id":2000010,"name":"Loyka","isActive":false,"title":"Loyka","metaTagDescription":"Camera, Photo","imageUrl":null},{"id":2000011,"name":"Kodak","isActive":false,"title":"Kodak","metaTagDescription":"Camera, Photo","imageUrl":null},{"id":2000012,"name":"Appel","isActive":false,"title":"Appel","metaTagDescription":"Computer, Notebook","imageUrl":null},{"id":2000013,"name":"Azer","isActive":false,"title":"Azer","metaTagDescription":"","imageUrl":null},{"id":2000014,"name":"Semsung","isActive":false,"title":"Semsung","metaTagDescription":"Mobile, Cellphone","imageUrl":null},{"id":2000015,"name":"Vauve","isActive":false,"title":"Vauve","metaTagDescription":"Video game, PC","imageUrl":null},{"id":2000016,"name":"Nentendo","isActive":false,"title":"Nentendo","metaTagDescription":"Video game, PC","imageUrl":null},{"id":2000017,"name":"Macrosoft","isActive":false,"title":"Macrosoft","metaTagDescription":"Video game, PC","imageUrl":null},{"id":2000018,"name":"Mottarola","isActive":false,"title":"Mottarola","metaTagDescription":"Mobile, Cellphone","imageUrl":null},{"id":2000019,"name":"Robot","isActive":false,"title":"Robot","metaTagDescription":"Meta tags, Robot","imageUrl":null},{"id":2000020,"name":"Lemovo","isActive":false,"title":"Lemovo","metaTagDescription":"Lemovo, PC","imageUrl":null},{"id":2000021,"name":"Dill","isActive":false,"title":"Dill","metaTagDescription":"Dell, Notebook","imageUrl":null},{"id":2000022,"name":"LB","isActive":false,"title":"LB","metaTagDescription":"Mobile, Cellphone","imageUrl":null},{"id":2000023,"name":"Fillips","isActive":false,"title":"Fillips","metaTagDescription":"TV, Television","imageUrl":null},{"id":2000024,"name":"Filco","isActive":false,"title":"Filco","metaTagDescription":"TV, Television","imageUrl":null},{"id":2000025,"name":"GBL","isActive":false,"title":"GBL","metaTagDescription":"TV, Television","imageUrl":null},{"id":2000026,"name":"Superlaser","isActive":false,"title":"Superlaser","metaTagDescription":"Superlaser, Audio","imageUrl":null},{"id":2000027,"name":"Beets","isActive":false,"title":"Beets","metaTagDescription":"Beets, Audio","imageUrl":null},{"id":2000028,"name":"Xaomi","isActive":false,"title":"Xaomi","metaTagDescription":"Xaomi, Mobile","imageUrl":null},{"id":2000029,"name":"Garman","isActive":false,"title":"Garman","metaTagDescription":"Garman, Mobile","imageUrl":null},{"id":2000030,"name":"PT-Link","isActive":false,"title":"PT-Link","metaTagDescription":"Router, Internet","imageUrl":null},{"id":2000031,"name":"Logitecho","isActive":false,"title":"Logitecho","metaTagDescription":"Logitecho, Camera","imageUrl":null},{"id":2000032,"name":"Air","isActive":false,"title":"Air","metaTagDescription":"Air, Audio","imageUrl":null},{"id":2000033,"name":"Blekin","isActive":false,"title":"Blekin","metaTagDescription":"Blekin, Audio","imageUrl":null},{"id":2000034,"name":"Incaser","isActive":false,"title":"Incaser","metaTagDescription":"Incaser, Audio","imageUrl":null}]`
  ),
}

export const catalogCategory3Fetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/category/tree/3',
  init: undefined,
  result: JSON.parse(
    '[{"id":9282,"name":"Office","hasChildren":true,"url":"http://storeframework.vtexcommercestable.com.br/office","children":[{"id":9295,"name":"Desks","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/office/desks","children":[],"Title":"Desks","MetaTagDescription":"Desks for better productivity"},{"id":9296,"name":"Chairs","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/office/chairs","children":[],"Title":"Chairs","MetaTagDescription":"Comfort chairs"}],"Title":"Office","MetaTagDescription":"For the office and home office"},{"id":9285,"name":"Kitchen and Home Appliances","hasChildren":true,"url":"http://storeframework.vtexcommercestable.com.br/kitchen-and-home-appliances","children":[{"id":9293,"name":"Fridges","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/kitchen-and-home-appliances/fridges","children":[],"Title":"Fridges","MetaTagDescription":"Fridges for the penguin"},{"id":9294,"name":"Appliances","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/kitchen-and-home-appliances/appliances","children":[],"Title":"Appliances","MetaTagDescription":"Appliances for you"}],"Title":"Home Appliances","MetaTagDescription":"Stay home with style"},{"id":9286,"name":"Computer and Software","hasChildren":true,"url":"http://storeframework.vtexcommercestable.com.br/computer-and-software","children":[{"id":9291,"name":"Smartphones","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/computer-and-software/smartphones","children":[],"Title":"Smartphones","MetaTagDescription":"You know what it is"},{"id":9292,"name":"Gadgets","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/computer-and-software/gadgets","children":[],"Title":"Gadgets","MetaTagDescription":"Gadgets for you"}],"Title":"Computer and Software","MetaTagDescription":"Get in touch with tomorrows world today"},{"id":9297,"name":"Technology","hasChildren":false,"url":"http://storeframework.vtexcommercestable.com.br/technology","children":[],"Title":"Technology","MetaTagDescription":"Technology"}]'
  ),
}

export const catalogPageTypeSkechers = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/skechers',
  init: undefined,
  result: JSON.parse(
    `{"id":"2000001","name":"Skechers","url":"storeframework.vtexcommercestable.com.br/Skechers","title":"Skechers","metaTagDescription":"Sport, casual, work, wide, kids' & performance shoes with style, comfort, innovation, quality & value.","pageType":"Brand"}`
  ),
}

export const catalogPageTypeAdidas = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/adidas',
  init: undefined,
  result: JSON.parse(
    `{"id":"2000004","name":"adidas","url":"storeframework.vtexcommercestable.com.br/adidas","title":"adidas","metaTagDescription":"adidas shoes, clothing and view new collections for adidas Originals, running, football, soccer, training and much more.","pageType":"Brand"}`
  ),
}

export const catalogPageTypeAcer = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/acer',
  init: undefined,
  result: JSON.parse(
    `{"id":"2000002","name":"Acer","url":"storeframework.vtexcommercestable.com.br/Acer","title":"Acer","metaTagDescription":"Acer laptops, desktops as well as servers and storage, personal digital assistance (PDA), peripherals, peripherals and e-business services for government, business, education, and home users.","pageType":"Brand"}`
  ),
}

export const catalogPageTypeIRobot = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/irobot',
  init: undefined,
  result: JSON.parse(
    `{"id":"2000003","name":"iRobot","url":"storeframework.vtexcommercestable.com.br/iRobot","title":"iRobot","metaTagDescription":"iRobot, the leading global consumer robot company, designs and builds robots that empower people to do more both inside and outside of the home.","pageType":"Brand"}`
  ),
}

export const catalogPageTypeBrand = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/brand',
  init: undefined,
  result: JSON.parse(
    `{"id":"9280","name":"Brand","url":"storeframework.vtexcommercestable.com.br/Brand","title":"Brand","metaTagDescription":"Brand","pageType":"Brand"}`
  ),
}
