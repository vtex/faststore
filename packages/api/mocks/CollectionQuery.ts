export const CollectionDesksQuery = `query CollectionQuery {
  collection(slug: "desks") {
    id
    slug
    type
    seo {
      title
      titleTemplate
      description
      canonical
    }
    breadcrumbList{
      itemListElement {
        item
        name
        position
      }
      numberOfItems
    }
    meta {
      selectedFacets {
        key
        value
      }
    }
  }
}
`

export const pageTypeDesksFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/desks',
  init: undefined,
  result: JSON.parse(
    '{"id":"9295","name":"Desks","url":"storeframework.vtexcommercestable.com.br/Office/Desks","title":"Desks","metaTagDescription":"Desks for better productivity","pageType":"Category"}'
  ),
}

export const pageTypeOfficeFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/office',
  init: undefined,
  result: JSON.parse(
    '{"id":"9282","name":"Office","url":"storeframework.vtexcommercestable.com.br/Office","title":"Office","metaTagDescription":"For the office and home office","pageType":"Department"}'
  ),
}

export const pageTypeOfficeDesksFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/portal/pagetype/office/desks',
  init: undefined,
  result: JSON.parse(
    '{"id":"9295","name":"Desks","url":"storeframework.vtexcommercestable.com.br/Office/Desks","title":"Desks","metaTagDescription":"Desks for better productivity","pageType":"Category"}'
  ),
}
