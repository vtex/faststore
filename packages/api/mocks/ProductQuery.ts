export const ProductByIdQuery = `query ProductQuery {
  product(locator: [{key: "id", value: "64953394"}]) {
    slug
    name
    productID
    description
    sku
    gtin
    seo {
      title
      titleTemplate
      description
      canonical
    }
    breadcrumbList {
      itemListElement {
        item
        name
        position
      }
    }
    brand {
      name
    }
    image {
      url
      alternateName
    }
    offers {
      highPrice
      lowPrice
      offerCount
      priceCurrency
    }
    review {
      reviewRating {
        ratingValue
        bestRating
      }
      author {
        name
      }
    }
    aggregateRating {
      ratingValue
      reviewCount
    }
    isVariantOf {
      hasVariant {
        name
        slug
        productID
      }
    }
    additionalProperty {
      value
      name
    }
  }
}
`

export const productSearchFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/io/_v/api/intelligent-search/product_search/trade-policy/1?page=1&count=1&query=sku%3A64953394&sort=&fuzzy=auto&locale=en-US&hideUnavailableItems=false',
  init: undefined,
  result: {
    products: [
      {
        cacheId: 'sp-29913569',
        productId: '29913569',
        description: 'Aut omnis nobis tenetur.',
        productName: 'Unbranded Concrete Table Small',
        productReference: '4715709796003',
        linkText: 'unbranded-concrete-table-small',
        brand: 'Brand',
        brandId: 9280,
        link: '/unbranded-concrete-table-small/p',
        categories: ['/Office/Desks/', '/Office/'],
        categoryId: '9295',
        categoriesIds: ['/9282/9295/', '/9282/'],
        priceRange: {
          sellingPrice: {
            highPrice: 200.64,
            lowPrice: 200.64,
          },
          listPrice: {
            highPrice: 297.7,
            lowPrice: 297.7,
          },
        },
        specificationGroups: [
          {
            originalName: 'allSpecifications',
            name: 'allSpecifications',
            specifications: [
              {
                originalName: 'sellerId',
                name: 'sellerId',
                values: ['1'],
              },
            ],
          },
        ],
        skuSpecifications: [],
        productClusters: [],
        clusterHighlights: [],
        properties: [
          {
            name: 'sellerId',
            originalName: 'sellerId',
            values: ['1'],
          },
        ],
        items: [
          {
            sellers: [
              {
                sellerId: '1',
                sellerName: 'VTEX',
                addToCartLink: '',
                sellerDefault: true,
                commertialOffer: {
                  DeliverySlaSamplesPerRegion: {},
                  DeliverySlaSamples: [],
                  AvailableQuantity: 10000,
                  discountHighlights: [],
                  Installments: [
                    {
                      Value: 200.64,
                      InterestRate: 0,
                      TotalValuePlusInterestRate: 200.64,
                      NumberOfInstallments: 1,
                      Name: 'Boleto Banc�rio � vista',
                      PaymentSystemName: 'Boleto Banc�rio',
                    },
                    {
                      Value: 200.64,
                      InterestRate: 0,
                      TotalValuePlusInterestRate: 200.64,
                      NumberOfInstallments: 1,
                      Name: 'Free � vista',
                      PaymentSystemName: 'Free',
                    },
                  ],
                  Price: 200.64,
                  ListPrice: 297.7,
                  spotPrice: 200.64,
                  taxPercentage: 0,
                  PriceWithoutDiscount: 200.64,
                  Tax: 0,
                  GiftSkuIds: [],
                  BuyTogether: [],
                  ItemMetadataAttachment: [],
                  RewardValue: 0,
                  PriceValidUntil: '2023-04-07T14:08:58Z',
                  GetInfoErrorMessage: null,
                  CacheVersionUsedToCallCheckout: '',
                  teasers: [],
                },
              },
            ],
            images: [
              {
                imageId: '186495',
                cacheId: '186495',
                imageTag: '',
                imageLabel: 'et',
                imageText: 'et',
                imageUrl:
                  'https://storeframework.vtexassets.com/arquivos/ids/186495/corporis.jpg?v=637755567185370000',
              },
              {
                imageId: '186492',
                cacheId: '186492',
                imageTag: '',
                imageLabel: 'in',
                imageText: 'in',
                imageUrl:
                  'https://storeframework.vtexassets.com/arquivos/ids/186492/qui.jpg?v=637755567174570000',
              },
              {
                imageId: '186493',
                cacheId: '186493',
                imageTag: '',
                imageLabel: 'consectetur',
                imageText: 'consectetur',
                imageUrl:
                  'https://storeframework.vtexassets.com/arquivos/ids/186493/possimus.jpg?v=637755567178470000',
              },
              {
                imageId: '186494',
                cacheId: '186494',
                imageTag: '',
                imageLabel: 'ea',
                imageText: 'ea',
                imageUrl:
                  'https://storeframework.vtexassets.com/arquivos/ids/186494/nihil.jpg?v=637755567181900000',
              },
            ],
            itemId: '64953394',
            name: 'fuchsia',
            nameComplete: 'Unbranded Concrete Table Small fuchsia',
            complementName:
              'Repellendus ipsum suscipit. Tempore consectetur illo dicta ducimus qui ut tempore. Consequatur non laboriosam aut deleniti doloribus nostrum ab et. Odio molestias hic dolor sunt ipsam non. Blanditiis rerum aut dolorum ratione eveniet voluptatibus. Laborum incidunt velit est est laudantium eos.',
            referenceId: [
              {
                Key: 'RefId',
                Value: '1346198062637',
              },
            ],
            measurementUnit: 'un',
            unitMultiplier: 1,
            variations: [],
            ean: '',
            modalType: '',
            videos: [],
            attachments: [],
            isKit: false,
          },
        ],
        origin: 'intelligent-search',
      },
    ],
    recordsFiltered: 1,
    correction: {
      misspelled: true,
    },
    fuzzy: 'auto',
    operator: 'and',
    translated: false,
    pagination: {
      count: 1,
      current: {
        index: 1,
        proxyUrl:
          'search/trade-policy/1?page=1&count=1&query=sku:64953394&sort=&fuzzy=auto&operator=and',
      },
      before: [],
      after: [],
      perPage: 1,
      next: {
        index: 0,
      },
      previous: {
        index: 0,
      },
      first: {
        index: 0,
      },
      last: {
        index: 0,
      },
    },
  },
}
