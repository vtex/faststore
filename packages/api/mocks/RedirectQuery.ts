export const RedirectQueryTermTech = `query RedirectSearch {
    redirect(term: "tech") {
        url
    }
  }
  `

export const redirectTermTechFetch = {
    info: 'https://storeframework.vtexcommercestable.com.br/api/io/_v/api/intelligent-search/product_search/trade-policy/1?page=2&count=1&query=tech&sort=&fuzzy=auto&locale=en-US&hideUnavailableItems=false',
    init: undefined,
    result: {
        products: [],
        recordsFiltered: 0,
        fuzzy: 'auto',
        operator: 'and',
        redirect: '/technology',
        translated: false,
        pagination: {
            count: 1,
            current: {
                index: 0,
            },
            before: [],
            after: [],
            perPage: 0,
            next: {
                index: 0,
            },
            previous: {
                index: 0,
            },
            first: {
                index: 0,
            },
            last: {
                index: 0,
            },
        },
    },
}
