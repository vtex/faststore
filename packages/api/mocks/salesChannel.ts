export const salesChannelStaleFetch = {
  info: 'https://storeframework.vtexcommercestable.com.br/api/catalog_system/pub/saleschannel/1',
  init: undefined,

  result: JSON.parse(
    '{"Id":1,"Name":"Principal","IsActive":true,"ProductClusterId":null,"CountryCode":"USA","CultureInfo":"en-US","TimeZone":"Central Standard Time","CurrencyCode":"USD","CurrencySymbol":"$","CurrencyLocale":1033,"CurrencyFormatInfo":{"CurrencyDecimalDigits":2,"CurrencyDecimalSeparator":".","CurrencyGroupSeparator":",","CurrencyGroupSize":3,"StartsWithCurrencySymbol":true},"Origin":null,"Position":1,"ConditionRule":null,"CurrencyDecimalDigits":null}'
  ),
}