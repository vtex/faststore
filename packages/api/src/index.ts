import { makeExecutableSchema } from '@graphql-tools/schema'

import {
  getContextFactory as getContextFactoryVTEX,
  getResolvers as getResolversVTEX,
} from './platforms/vtex'
import { typeDefs } from './typeDefs'
import type { Options as OptionsVTEX } from './platforms/vtex'

export * from './__generated__/schema'

export type Options = OptionsVTEX

const platforms = {
  vtex: {
    getResolvers: getResolversVTEX,
    getContextFactory: getContextFactoryVTEX,
  },
}

export const getTypeDefs = () => typeDefs

export const getResolvers = (options: Options) =>
  platforms[options.platform].getResolvers(options)

export const getContextFactory = (options: Options) =>
  platforms[options.platform].getContextFactory(options)

export const getSchema = async (options: Options) =>
  makeExecutableSchema({
    resolvers: getResolvers(options),
    typeDefs: getTypeDefs(),
  })
