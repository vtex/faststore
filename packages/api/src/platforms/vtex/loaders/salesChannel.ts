import DataLoader from "dataloader";

import { Options } from "..";
import { Clients } from "../clients";

import type { SalesChannel } from "./../clients/commerce/types/SalesChannel";

export const getSalesChannelLoader = (_: Options, clients: Clients) => {
  const loader = async (channels: readonly string[]) =>
    Promise.all(
      channels.map((sc) => clients.commerce.catalog.salesChannel(sc)),
    );

  return new DataLoader<string, SalesChannel>(loader);
};
