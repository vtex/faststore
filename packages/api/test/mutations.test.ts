import { execute, parse } from 'graphql'

import {
  checkoutOrderFormCustomDataInvalidFetch,
  checkoutOrderFormCustomDataStaleFetch,
  checkoutOrderFormCustomDataValidFetch,
  checkoutOrderFormInvalidFetch,
  checkoutOrderFormItemsInvalidFetch,
  checkoutOrderFormItemsValidFetch,
  checkoutOrderFormStaleFetch,
  checkoutOrderFormValidFetch,
  InvalidCart,
  productSearchPage1Count1Fetch,
  ValidateCartMutation,
  ValidCart,
} from '../mocks/ValidateCartMutation'
import { getContextFactory, getSchema } from '../src'
import { salesChannelStaleFetch } from '../mocks/salesChannel'
import type { Options } from '../src'

const apiOptions = {
  platform: 'vtex',
  account: 'storeframework',
  environment: 'vtexcommercestable',
  channel: '{"salesChannel":"1"}',
  locale: 'en-US',
  hideUnavailableItems: false,
  incrementAddress: false,
  flags: {
    enableOrderFormSync: true,
  },
} as Options

const mockedFetch = jest.fn()

const createRunner = () => {
  const schemaPromise = getSchema(apiOptions)
  const contextFactory = getContextFactory(apiOptions)

  return async (query: string, variables?: any) => {
    const schema = await schemaPromise
    const context = contextFactory({})

    return execute(
      schema,
      parse(query),
      null,
      { ...context, headers: { cookie: '' } },
      variables
    )
  }
}

function pickFetchAPICallResult(
  info: RequestInfo,
  _: RequestInit | undefined,
  expectedFetchAPICalls: Array<Record<'info' | 'init' | 'result', unknown>>
) {
  for (const call of expectedFetchAPICalls) {
    if (info === call.info) {
      return call.result
    }
  }

  throw new Error(
    `fetchAPI was called with an unexpected 'info' argument.\ninfo: ${info}`
  )
}

jest.mock('../src/platforms/vtex/clients/fetch.ts', () => ({
  fetchAPI: async (info: RequestInfo, init?: RequestInit) =>
    mockedFetch(info, init),
}))

const run = createRunner()

// Always clear the mocked fetch before each test so we can count and validate
// the calls performed by each query independently.
beforeEach(() => {
  mockedFetch.mockClear()
})

test('`validateCart` mutation should return `null` when a valid cart is passed', async () => {
  const fetchAPICalls = [
    checkoutOrderFormValidFetch,
    checkoutOrderFormItemsValidFetch,
    checkoutOrderFormCustomDataValidFetch,
  ]

  mockedFetch.mockImplementation((info, init) =>
    pickFetchAPICallResult(info, init, fetchAPICalls)
  )

  const response = await run(ValidateCartMutation, { cart: ValidCart })

  expect(mockedFetch).toHaveBeenCalledTimes(3)

  fetchAPICalls.forEach((fetchAPICall) => {
    expect(mockedFetch).toHaveBeenCalledWith(
      fetchAPICall.info,
      fetchAPICall.init
    )
  })

  expect(response).toEqual({ data: { validateCart: null } })
})

test('`validateCart` mutation should return the full order when an invalid cart is passed', async () => {
  const fetchAPICalls = [
    checkoutOrderFormInvalidFetch,
    checkoutOrderFormItemsInvalidFetch,
    checkoutOrderFormCustomDataInvalidFetch,
    productSearchPage1Count1Fetch,
    salesChannelStaleFetch,
  ]

  mockedFetch.mockImplementation((info, init) =>
    pickFetchAPICallResult(info, init, fetchAPICalls)
  )

  const response = await run(ValidateCartMutation, { cart: InvalidCart })

  expect(mockedFetch).toHaveBeenCalledTimes(5)

  fetchAPICalls.forEach((fetchAPICall, index) => {
    expect(mockedFetch).toHaveBeenNthCalledWith(
      index + 1,
      fetchAPICall.info,
      fetchAPICall.init
    )
  })

  expect(response).toMatchSnapshot()
})

test('`validateCart` mutation should return new cart when etag is stale', async () => {
  const fetchAPICalls = [
    checkoutOrderFormStaleFetch,
    checkoutOrderFormCustomDataStaleFetch,
    productSearchPage1Count1Fetch,
    salesChannelStaleFetch,
  ]

  mockedFetch.mockImplementation((info, init) =>
    pickFetchAPICallResult(info, init, fetchAPICalls)
  )

  const response = await run(ValidateCartMutation, { cart: InvalidCart })

  expect(mockedFetch).toHaveBeenCalledTimes(4)

  fetchAPICalls.forEach((fetchAPICall, index) => {
    expect(mockedFetch).toHaveBeenNthCalledWith(
      index + 1,
      fetchAPICall.info,
      fetchAPICall.init
    )
  })

  expect(response).toMatchSnapshot()
})
