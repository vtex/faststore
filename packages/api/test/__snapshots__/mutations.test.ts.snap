// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`validateCart\` mutation should return new cart when etag is stale 1`] = `
Object {
  "data": Object {
    "validateCart": Object {
      "messages": Array [],
      "order": Object {
        "acceptedOffer": Array [
          Object {
            "availability": "https://schema.org/InStock",
            "itemCondition": "https://schema.org/NewCondition",
            "itemOffered": Object {
              "name": "plum",
            },
            "listPrice": 557.57,
            "price": 349.12,
            "priceCurrency": "USD",
            "quantity": 2,
            "seller": Object {
              "identifier": "1",
            },
            "sellingPrice": 349.12,
          },
        ],
        "orderNumber": "edbe3b03c8c94827a37ec5a6a4648fd2",
      },
    },
  },
}
`;

exports[`\`validateCart\` mutation should return the full order when an invalid cart is passed 1`] = `
Object {
  "data": Object {
    "validateCart": Object {
      "messages": Array [],
      "order": Object {
        "acceptedOffer": Array [
          Object {
            "availability": "https://schema.org/InStock",
            "itemCondition": "https://schema.org/NewCondition",
            "itemOffered": Object {
              "name": "plum",
            },
            "listPrice": 557.57,
            "price": 349.12,
            "priceCurrency": "USD",
            "quantity": 2,
            "seller": Object {
              "identifier": "1",
            },
            "sellingPrice": 349.12,
          },
        ],
        "orderNumber": "edbe3b03c8c94827a37ec5a6a4648fd2",
      },
    },
  },
}
`;
